prov/rxd/src/rxd_ep.c

```diff
@@ -885,16 +912,39 @@ static void rxd_peer_timeout(struct rxd_ep *rxd_ep, struct rxd_peer *peer)
 	     	peer->unacked_cnt--;
 	}
 
+	dlist_remove(&peer->entry);
+}
+
+static void rxd_progress_pkt_list(struct rxd_ep *ep, struct dlist_entry *list)
+{
+	struct rxd_pkt_entry *pkt_entry;
+	struct dlist_entry *tmp;
+	uint64_t current;
+	int ret;
+
+	current = fi_gettime_ms();
+	dlist_foreach_container_safe(list, struct rxd_pkt_entry,
+				     pkt_entry, d_entry, tmp) {
+		if (current < pkt_entry->retry_time)
+			continue;
+
+		if (pkt_entry->retry_cnt > RXD_MAX_PKT_RETRY) {
+			rxd_peer_timeout(ep,
+				&ep->peers[rxd_get_base_hdr(pkt_entry)->peer]);
+			break;
```

Comment:
Why do we break here, rather than continuing to check to rest of the packets in the list?

Reply:
Progress is made per peer so if the peer times out, it removes the unacked packets and completes the pending entries in error if the peer becomes unresponsive. So there's not much of a point in continuing to progress the packets.
And same for the next comment. If we have an issue sending a packet, there's not much of a point to try the rest since they will probably also result in an error or, if they do get sent, they'll be out of order at the receive side.

Reply:
got it -thanks